### PR TITLE
Pass serialization scope to lazy loaders

### PR DESCRIPTION
**Pull request by Bajena**
It doesn't necessarily have to get to master, but I've heard some voices that it might be useful to have access to the serialization scope.

closes https://github.com/Bajena/ams_lazy_relationships/issues/41

### CODE REVIEW

File: lib/ams_lazy_relationships/core/evaluation.rb [hunk 1]
```diff
@@ -13,51 +13,55 @@ module Evaluation
...
       end
     end
 
-    def deep_load_for_yielded_record(batch_record, lrm, level)
+    def deep_load_for_yielded_record(batch_record, scope, lrm, level)
       serializer = serializer_for(batch_record, lrm.reflection.options)
       return unless serializer
 
-      serializer.send(:load_all_lazy_relationships, batch_record, level + 1)
+      serializer.send(:load_all_lazy_relationships, batch_record, scope, level + 1)
```
Comment (stokarenko) [inline lib/ams_lazy_relationships/core/evaluation.rb]: Probably we need to take scope from nested serializer, something like

```suggestion
      serializer.send(:load_all_lazy_relationships, batch_record, serializer.scope, level + 1)
```

File: spec/core_spec.rb [hunk 2]
```diff
@@ -686,4 +686,89 @@ class CategorySerializer < BaseTestSerializer
...
   end
+
+  describe "using serialization scope" do
+    class BlogPostsLoader < AmsLazyRelationships::Loaders::Base
+      def load_data(records, loader, scope)
+        data = []
+
+        records.each do |r|
+          d = r.blog_posts
+          d = d.where(title: scope[:title]) if scope
```
Comment (stokarenko) [inline spec/core_spec.rb]: I'm not very familiar with Serializer scopes at the moment, but looks like such kind of customized loader will play against idempotency (serializer with lazy_relationships will produce different output than pure serializer), nor will duplicate the business logic (how scope-basic logic has been applied to pure serializer, why that logic was not applied to lazy serializer out-from-the-box?)..

File: lib/ams_lazy_relationships/core/evaluation.rb (hunk 1 above)
Comment (Bajena) [reply lib/ams_lazy_relationships/core/evaluation.rb]: I don't think that's possible, because AFAIK `serializer` here is the class not the instance.
Btw the scope is in most cases (I haven't seen other :P) provided by the Rails controller by calling e.g. `serialization_scope :current_user`, so this should be fine.

File: spec/core_spec.rb (hunk 2 above)
Comment (Bajena) [reply spec/core_spec.rb]: > serializer with lazy_relationships will produce different output than pure serializer

Not necessarily - imagine you had a following use case (it's stupid I know, but couldn't imagine better ATM :D):
```ruby
class BlogPostsController
  serialization_scope :current_user

  def index
     render json: BlogPost.all
  end
end

class BlogPostSerializer
  has_many :comments do
    current_user.premium? ? object.comments : object.comments.first
  end
end
```

if you want to rework it using lazy relationships you could do sth like:
```ruby
class BlogPostSerializer
  lazy_relationship :comments

  has_many :comments do
    # We still need to load ALL comments for non-premium users
    current_user.premium? ? lazy_comments : lazy_comments.first
  end
end
```

with scope passed to lazy loaders you can do:
```ruby
  class CommentsLoader < AmsLazyRelationships::Loaders::Base
      def load_data(records, loader, current_user)
        # load all comments for premium users, but only one for non-premium
      end
  end

class BlogPostSerializer
  lazy_has_many :comments, loader: CommentsLoader
end
```

Do you get the idea now?

Comment (stokarenko) [reply spec/core_spec.rb]: @Bajena got the idea, need some time to think about )

**Comment by Bajena** [comment]
@stokarenko  I'll move on with this PR as soon as your PRs are shipped :)

**Comment by Bajena** [comment]
@stokarenko I've rebased to master and added a section about custom loaders to README. Take a look if you have some time :)

File: spec/core_spec.rb (hunk 2 above)
Comment (stokarenko) [reply spec/core_spec.rb]: The last concern that I feel is that such kind of conditional filtering is out from serializer's scope of responsibilities, as well as it is out from JSON API specification. Even more, each step deeper than straightforward rendering of plain AR association within serializer's `includes` will lead the team to the dreams about include-less serializers, eventually.

The need to filter association sounds so close to the need to paginate association, isn't? There is an example [how that looks at the end](https://discuss.jsonapi.org/t/paginating-relationships-includes/820/3)

> TLDR;
> We did not implement includes.
>
> After theory-crafting out scaling issues, it became apparent we would need to paginate includes if we returned them… BUT we would also want to have a mechanism for allowing to specify what “page” we are paginating each includes on and return them properly. This was a pain in the … well you know. It also came to our attention that if we did paginate the includes, some partners may assume the includes we would returning are all of them, not just a subset.
> 
> Because the spec does no specify how to do this completely, we decided not to implement includes. (includes are not required for spec as far as I can tell). This does mean partners must make more requests. But the up side: we can limit the number of queries per request to a manageable number and keep response times sub 100ms.

In general I agree that controller scope passed into serializer can help to deal with very special business cases, but that can be a marker of unlucky API design in the same time, leading to the similar `We did not implement includes` after the days..

File: spec/core_spec.rb [hunk 3]
```diff
@@ -1018,4 +1018,89 @@ class User3Serializer < BaseTestSerializer
...
+        data.flatten.compact.uniq
+      end
+
+      def batch_key(_)
+        "Key"
+      end
+    end
+
+    let(:level0_serializer_class) do
+      class Level1Serializer11 < BaseTestSerializer
```
Comment (stokarenko) [inline spec/core_spec.rb]: That should be `Level1Serializer16` I believe )
I think will be good some day to refactor all the tests there, in order to isolate them better )
